src/java.base/share/classes/jdk/internal/lang/stable/StableValueImpl.java

```diff
@@ -0,0 +1,204 @@
+ *
+ * @param <T> type of the content
+ */
+public final class StableValueImpl<T> implements StableValue<T> {
+
+    // Unsafe allows StableValue to be used early in the boot sequence
+    static final Unsafe UNSAFE = Unsafe.getUnsafe();
+
+    // Unsafe offsets for direct field access
+    private static final long CONTENT_OFFSET =
+            UNSAFE.objectFieldOffset(StableValueImpl.class, "content");
+
+    // Generally, fields annotated with `@Stable` are accessed by the JVM using special
+    // memory semantics rules (see `parse.hpp` and `parse(1|2|3).cpp`).
+    //
+    // This field is used directly and reflectively via Unsafe using explicit memory semantics.
+    //
+    // | Value          |  Meaning      |
+    // | -------------- |  ------------ |
+    // | null           |  Unset        |
+    // | nullSentinel() |  Set(null)    |
+    // | other          |  Set(other)   |
+    //
+    @Stable
+    private Object content;
+
+    // Only allow creation via the factory `StableValueImpl::newInstance`
+    private StableValueImpl() {}
+
+    @ForceInline
+    @Override
+    public boolean trySet(T value) {
+        if (wrappedContentAcquire() != null) {
+            return false;
+        }
+        // Prevent reentry via an orElseSet(supplier)
+        if (Thread.holdsLock(this)) {
+            throw new IllegalStateException("Recursing supplier detected");
+        }
+        // Mutual exclusion is required here as `orElseSet` might also
+        // attempt to modify the `wrappedValue`
+        synchronized (this) {
+            return wrapAndSet(value);
+        }
+    }
+
+    @ForceInline
+    @Override
+    public void setOrThrow(T value) {
+        if (!trySet(value)) {
+            throw new IllegalStateException("Cannot set the content to " + value +
+                    " because the content is already set: " + orElseThrow());
+        }
+    }
+
+    @ForceInline
+    @Override
+    public T orElseThrow() {
+        final Object t = wrappedContentAcquire();
+        if (t == null) {
+            throw new NoSuchElementException("No content set");
+        }
+        return unwrap(t);
+    }
+
+    @ForceInline
+    @Override
+    public T orElse(T other) {
+        final Object t = wrappedContentAcquire();
+        return (t == null) ? other : unwrap(t);
+    }
+
+    @ForceInline
+    @Override
+    public boolean isSet() {
+        return wrappedContentAcquire() != null;
+    }
+
+    @ForceInline
+    @Override
+    public T orElseSet(Supplier<? extends T> supplier) {
+        Objects.requireNonNull(supplier);
+        final Object t = wrappedContentAcquire();
+        return (t == null) ? orElseSetSlowPath(supplier) : unwrap(t);
+    }
+
+    @DontInline
+    private T orElseSetSlowPath(Supplier<? extends T> supplier) {
+        // Prevent reentry
+        if (Thread.holdsLock(this)) {
+            throw new IllegalStateException("Recursing supplier detected: " + supplier);
```

Comment:
Is it right to include `supplier` in the message? The actual recursing supplier could be another one:

```java
var s = StableValue.<Integer>of();
// throws java.lang.IllegalStateException: Recursing supplier detected: supplier 2
s.orElseSet(new Supplier<>() {
    public String toString() {return "supplier 1";}
    public Integer get() {
        s.orElseSet(new Supplier<>() {
            public String toString() {return "supplier 2";}
            public Integer get() {return 2;}
        });
        return 1;
    }
});
```
The exception message could also be confusing in cases like this:
```java
var s = StableValue.of();
synchronized (s) {
    // throws java.lang.IllegalStateException: Recursing supplier detected
    s.trySet(null);
}
```